### PR TITLE
New Taxonomies Integration

### PR DESCRIPTION
**Pull request by lmcnulty**
[Deploy Preview](https://deploy-preview-18--lmcnulty-staging.netlify.app)  
[Dev PR](https://github.com/lmcnulty/aiid/pull/18)

Preview:

- [/taxonomies](https://deploy-preview-18--lmcnulty-staging.netlify.app/taxonomies/)
- [/taxonomies/gmf](https://deploy-preview-18--lmcnulty-staging.netlify.app/taxonomy/gmf/)
- [discover with GMF classification filter](https://deploy-preview-18--lmcnulty-staging.netlify.app/apps/discover/?classifications=GMF%3AKnown%20AI%20Goal%3AAutonomous%20Driving&display=details&is_incident_report=true&page=1)
- [/summaries/spatial/](https://deploy-preview-18--lmcnulty-staging.netlify.app/summaries/spatial/) (with GMF classifications)
- [Edit GMF](https://deploy-preview-18--lmcnulty-staging.netlify.app/cite/2/?edit_taxonomy=GMF) (logged in)
- [Edit CSETv1](https://deploy-preview-18--lmcnulty-staging.netlify.app/cite/2/?edit_taxonomy=CSETv1) (logged in)
- [Citation page with GMF](https://deploy-preview-18--lmcnulty-staging.netlify.app/cite/10/) (logged out)

---

https://user-images.githubusercontent.com/25443411/216205757-078544fa-100a-470c-adc6-5aa4640e4a80.mp4

Setting `GATSBY_CPU_COUNT` to 1 or 2 seems to let the build go through.

<details>
<summary>Former Build issues</summary>

I'm trying to make a deploy preview [from my fork](https://github.com/lmcnulty/aiid/pull/18), but [the build](https://app.netlify.com/sites/lmcnulty-staging/deploys/63d7e862d521873ab50ccb5b) is failing with 

```
Error message
Command was killed with SIGKILL (Forced termination): gatsby build
 ​
Error location
In Build command from Netlify app:
gatsby build
```

This is the same problem from https://github.com/responsible-ai-collaborative/aiid/issues/953, but I believe my build is running on the organization's plan.

</details>

---

Per #1497.

> - [x] Remove the resources taxonomy entirely. We are not currently using it and it is better to re-introduce it after the CSETv2 and GMF taxonomies are integrated. This should be accomplished simply by removing the resources document from the taxa collection and removing the resources collection.
> - [x] Copy the CSETv1 taxa document to a new schema in the taxa collection.
> - [x] Update the schema to match the [v2 specification](https://docs.google.com/document/d/1SYJfiP_E_gya89ljxZumC9z6MkP4z1I-bfgnDs6kIhQ/edit#heading=h.ehpbfkrxkvv)
> - [x] Update the taxonomy feature, which formerly put classifications for the CSET taxonomy into "classifications" and classifications for the resources taxonomy into "resources" so that all taxonomies place their classifications into the "classifications" collection. Taxonomies should be fully specified in the taxa collection and its data should be in the classifications collection.
> - [x] The CSET v2 taxonomy should then be displayed on the citation pages and it should be possible to introduce new classifications which takes the taxonomy from hidden, to displayed.

<details>
<summary>Old comments on CSET taxonomy.</summary>
Currently adds a CSET2 taxa entry that's the same as CSET except for the name. As far as updating it goes, it's not clear to me how the field table on the current draft of the annotation guide should correspond to a taxa entry in the database. Just looking at the top two rows, we have:

Field number | Field name | Description for annotators | Public field description (on incidentdatabase.ai) | Example | Notes
-- | -- | -- | -- | -- | --
1.1 | Incident number | The number of the incident in the AI Incident Database. | id | 1 |  
1.2 | Who is responsible for this row? | The full name of the annotator for this incident. | An ID designating the individual who classified this incident according to the CSET taxonomy | Heather Frase |  

"Who is responsible for this row" doesn't seem like a field name the same way "Incident number" does. Likewise, "id" doesn't seem like a description the same as "An ID designating…", and they couldn't really be used for the same purposes. For example, the former could be used as a table heading, whereas the latter could not.
</details>

## Testing / Breakage Attempts

<details>
<summary>Expand/collapse</summary>

- When logged in, shows CSET2 Taxonomy classifications

  image

- Doesn't show the CSET2 classifications when not logged in.

  image

- The edit form for CSET2 shows up
  
  image

- You can submit the edit form

  image

- The result appears in classifications collection under the CSET2 namespace
  
  image

- The added classifications appear on the citation page

  **Found issue**:
  image

  after restart server, CSET2 appears but CSET doesn't:
  
  image

  GraphQL mutation needs selection.

  Now shows both in the db but only CSET in the UI

  image

  image

  Fixed -- had to switch classifications query from `mongodbAiidprodClassifications` to `allMongodbAiidprodClassifications`

</details>

### CODE REVIEW

File: site/gatsby-site/src/templates/cite.js
Comment (cesarvarela) [inline site/gatsby-site/src/templates/cite.js:158]: It is better to avoid direct DOM access, so maybe use a `ref`?

File: site/gatsby-site/src/utils/AlgoliaUpdater.js
Comment (cesarvarela) [inline site/gatsby-site/src/utils/AlgoliaUpdater.js:175]: Same here. We probably want to crash the entire build if something throws.

File: site/gatsby-site/src/utils/AlgoliaUpdater.js
Comment (cesarvarela) [inline site/gatsby-site/src/utils/AlgoliaUpdater.js:189]: ditto

File: site/gatsby-site/src/utils/AlgoliaUpdater.js
Comment (cesarvarela) [inline site/gatsby-site/src/utils/AlgoliaUpdater.js:197]: Ok, I'm seeing this many times. Is it only for testing purposes?

**Review by cesarvarela — COMMENTED** [review]
- don't worry much about the `down` migrations because is tricky to get right, and it will be easier just to restore the DB (we could do a mongodump before merging)

- I'm aware of the deadline, so don't stress too much about my comments, but don't forget about the missing awaits and making sure we are not masking any critical error that should crash the build

**Comment by lmcnulty** [comment]
> The GMF taxonomy needs to have each of the classifications treated separately as a list

Is this just changing the classifications input from a string to a list? That's working on the [deploy preview](https://deploy-preview-18--lmcnulty-staging.netlify.app/cite/1/?edit_taxonomy=GMF) now.

image

**Comment by smcgregor** [comment]
Yes, that should be it.

File: site/gatsby-site/src/components/taxa/TaxonomyForm.js
Comment (lmcnulty) [reply site/gatsby-site/src/components/taxa/TaxonomyForm.js:38]: I found that when two taxonomies were open for editing at the same time, changing responses for one would sometimes change them for the other as well. To eliminate that class of errors, I set it so that only one taxonomy form could be rendered at a time. However, that means that that if you enter some values for taxonomy A, then hit edit to look at taxonomy B, when you return to taxonomy A, the values you entered are no longer there. To fix that, I needed to make it so the initial values are always set to whatever the most recently entered values were. That has to be debounced, though, because otherwise it makes typing laggy.

File: site/gatsby-site/migrations/2023.02.07T18.26.40.add-gmf-annotations.js
Comment (lmcnulty) [reply site/gatsby-site/migrations/2023.02.07T18.26.40.add-gmf-annotations.js:1699]: Hmm, those were in the [annotations JSON file that Nikiforos gave us](https://github.com/smcgregor/aiid-taxonomies/blob/master/annotations.json), but they probably don't belong. We should probably ask what those are supposed to be.

File: site/gatsby-site/src/components/taxa/TaxonomyForm.js
Comment (cesarvarela) [reply site/gatsby-site/src/components/taxa/TaxonomyForm.js:38]: That's really weird! Looks like this might cause a race condition if whatever you're waiting for takes longer than `500`. You can try with `0`, which is a bit less race-condition-prone. 

But eventually, we should fix it properly, which I think will require moving some `fetch` and `useEffect` around.

**Review by cesarvarela — APPROVED** [review]
Overall this looks good, except for the `debounce`d line. 

@smcgregor Should I merge to staging?

**Comment by smcgregor** [comment]
I am good to merge if @lmcnulty is. I think we will have a few more change sets flowing through on this, but we can do those iteratively after the migrations run and we live in the new more configurable world.

**Comment by smcgregor** [comment]
@lmcnulty messaging updates for  https://deploy-preview-18--lmcnulty-staging.netlify.app/taxonomy/gmf/

`The Goals, Methods, and Failures (GMF) taxonomy is a failure cause analysis taxonomy interrelating the goals of the system deployment, the system's methods, and their likely failings. Details on the process are available in the [recent work](https://arxiv.org/abs/2211.07280) published for the [SafeAI](https://safeai.webs.upv.es/) conference. [Comments and contributions](/contact/) on this in-development taxonomy are invited`

The cards on the page that have no classifications at present should have filler content indicating that nothing has been assigned a value for the card yet. Other changes you are making could obviate this suggested change.

**Comment by lmcnulty** [comment]
@smcgregor Does this look OK as a placeholder?



I think I'm good to merge when the build for the last commit finishes: https://app.netlify.com/sites/lmcnulty-staging/deploys/63e6d1700898b900080d36cb

**Comment by smcgregor** [comment]
Looks great. Go for merge when it completes.

**Comment by smcgregor** [comment]
@lmcnulty can you migrate anything that is still an open issue with the PR to new issues on GitHub? Also, please take a look at the GMF taxonomy as rendered on this page: https://deploy-preview-18--lmcnulty-staging.netlify.app/cite/72/

It features prominently in the presentation being delivered on Monday and there are non-blocking formatting issues to work through.

**Pull request merged**
